Adding UniformStreamByteDistributor

nmittler opened the pull request:
Motivation:

The current priority algorithm can yield poor per-stream goodput when either the number of streams is high or the connection window is small. When all priorities are the same (i.e. priority is disabled), we should be able to do better.

Modifications:

Added a new UniformStreamByteDistributor that ignores priority entirely and manages a queue of streams.  Each stream is allocated a minimum of 1KiB on each iteration.

Result:

Improved goodput when priority is not used.

nmittler commented:
@Scottmitch @ejona86 PTAL

nmittler commented:
Benchmark results:

```
Benchmark                                             (algorithm)  (connectionWindow)  (numStreams)   Mode  Cnt          Score            Error  Units
NoPriorityByteDistributionBenchmark.write                PRIORITY                1024           100  thrpt   10      26955.054 ±      12783.264  ops/s
NoPriorityByteDistributionBenchmark.write:minBytes       PRIORITY                1024           100  thrpt   10          9.156 ±          3.886  ops/s
NoPriorityByteDistributionBenchmark.write:totalBytes     PRIORITY                1024           100  thrpt   10   27601975.474 ±   13090062.288  ops/s
NoPriorityByteDistributionBenchmark.write                PRIORITY                1024         10000  thrpt   10       1333.529 ±         83.245  ops/s
NoPriorityByteDistributionBenchmark.write:minBytes       PRIORITY                1024         10000  thrpt   10          0.997 ±          0.002  ops/s
NoPriorityByteDistributionBenchmark.write:totalBytes     PRIORITY                1024         10000  thrpt   10    1365533.709 ±      85242.545  ops/s
NoPriorityByteDistributionBenchmark.write                PRIORITY               65536           100  thrpt   10      27388.701 ±      13896.645  ops/s
NoPriorityByteDistributionBenchmark.write:minBytes       PRIORITY               65536           100  thrpt   10        600.110 ±        254.905  ops/s
NoPriorityByteDistributionBenchmark.write:totalBytes     PRIORITY               65536           100  thrpt   10  509681719.118 ± 2844067724.871  ops/s
NoPriorityByteDistributionBenchmark.write                PRIORITY               65536         10000  thrpt   10        441.493 ±         28.712  ops/s
NoPriorityByteDistributionBenchmark.write:minBytes       PRIORITY               65536         10000  thrpt   10          5.976 ±          0.018  ops/s
NoPriorityByteDistributionBenchmark.write:totalBytes     PRIORITY               65536         10000  thrpt   10   28933654.546 ±    1881665.804  ops/s
NoPriorityByteDistributionBenchmark.write                 UNIFORM                1024           100  thrpt   10      21965.377 ±      40332.042  ops/s
NoPriorityByteDistributionBenchmark.write:minBytes        UNIFORM                1024           100  thrpt   10        582.481 ±        503.257  ops/s
NoPriorityByteDistributionBenchmark.write:totalBytes      UNIFORM                1024           100  thrpt   10   22492546.189 ±   41300010.688  ops/s
NoPriorityByteDistributionBenchmark.write                 UNIFORM                1024         10000  thrpt   10       1724.283 ±        112.897  ops/s
NoPriorityByteDistributionBenchmark.write:minBytes        UNIFORM                1024         10000  thrpt   10       1020.645 ±          2.563  ops/s
NoPriorityByteDistributionBenchmark.write:totalBytes      UNIFORM                1024         10000  thrpt   10    1765665.835 ±     115607.023  ops/s
NoPriorityByteDistributionBenchmark.write                 UNIFORM               65536           100  thrpt   10      21310.359 ±      39671.065  ops/s
NoPriorityByteDistributionBenchmark.write:minBytes        UNIFORM               65536           100  thrpt   10      38065.460 ±      32063.777  ops/s
NoPriorityByteDistributionBenchmark.write:totalBytes      UNIFORM               65536           100  thrpt   10   13525431.324 ±  657784010.009  ops/s
NoPriorityByteDistributionBenchmark.write                 UNIFORM               65536         10000  thrpt   10       1671.319 ±         87.628  ops/s
NoPriorityByteDistributionBenchmark.write:minBytes        UNIFORM               65536         10000  thrpt   10      65321.118 ±        177.599  ops/s
NoPriorityByteDistributionBenchmark.write:totalBytes      UNIFORM               65536         10000  thrpt   10  109531588.729 ±    5742820.574  ops/s
```

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,252 @@
+        long unallocated = totalStreamableBytes;
+        int allocationChunk = max(minAllocationChunk, (int)(unallocated / queue.size()));
+        while (maxBytes > 0 && !queue.isEmpty()) {
+            //int allocationChunk = max(minAllocationChunk, (int)(unallocated / queue.size()));
```
remove ?

nmittler replied:
done.

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,252 @@
+        @Override
+        void enqueue(State state) {
+            queue.addFirst(state);
+            state.enqueued = true;
```
remove the two lines and jus call `super.enqueue(...)`

nmittler replied:
They're different. The first time it's enqueued, it is added to the head of the queue ... after that it's added to the end.

normanmaurer replied:
gotcha

normanmaurer commented on microbench/src/main/java/io/netty/microbench/http2/NoPriorityByteDistributionBenchmark.java:
```diff
@@ -0,0 +1,315 @@
+    private static ByteCounter byteCounter;
+
+    /**
+     * Use an AuxCounter so we can measure the goodput.
```
s/goodput/throughput/ ?

nmittler replied:
The real goal here is goodput, which is why I'm specifically capturing the minimum per stream.

normanmaurer replied:
ok :)

normanmaurer commented on microbench/src/main/java/io/netty/microbench/http2/NoPriorityByteDistributionBenchmark.java:
```diff
@@ -0,0 +1,315 @@
+        }
+    }
+
+    private static class TestChannel extends AbstractChannel {
```
why not just mock it ?

nmittler replied:
Mocking channel ends up being a pain with all of the current logic in `DefaultHttp2RemoteFlowController`. It uses a lot of the interface and we've moved away from mocking it in http2 for that reason.

normanmaurer replied:
ok..

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,252 @@
+     * The remote flow control state for a single stream.
+     */
+    private final class State {
+        final Http2Stream stream;
```
Is this used?

nmittler replied:
No it's not ... good catch.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,252 @@
+            }
+        }
+
+        boolean hasFrame() {
```
Is this used anywhere?

nmittler replied:
no ... removed.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,252 @@
+            //int allocationChunk = max(minAllocationChunk, (int)(unallocated / queue.size()));
+
+            // Remove the head of the queue.
+            State state = queue.remove();
```
`remove()` -> `removeFirst()`?

We are being very particular about where we insert from using addLast and addFirst, perhaps we should do the same with removal?

nmittler replied:
done.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,252 @@
+            state.allocate(chunk);
+            maxBytes -= chunk;
+            allocated += chunk;
+            unallocated -= chunk;
```
Is keeping track of `unallocated` necessary? Seems like we could just use `totalStreamableBytes` above?

```java
int allocationChunk = max(minAllocationChunk, (int)(totalStreamableBytes / queue.size()));
```

nmittler replied:
No it's no longer necessary .... artifact of some previous experimentation.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,252 @@
+            int chunk = min(allocationChunk, min(maxBytes, state.unallocated()));
+            state.allocate(chunk);
+            maxBytes -= chunk;
+            allocated += chunk;
```
Consider reducing the number of operations (1 increment and 1 assign)  by removing this line and replacing with something like the following:

```java
final int maxBytesBefore = maxBytes;
while (...) {
...
}
return maxBytesBefore - maxBytes;
```

Scottmitch replied:
@nmittler - wdyt?

nmittler replied:
done.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,252 @@
+            allocated += chunk;
+            unallocated -= chunk;
+
+            if (state.unallocated() > 0) {
```
What if the state has empty frames? Do we need to use `hasFrame()` here instead?

nmittler replied:
Empty frames do not matter WRT allocation, they only matter when writing.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,252 @@
+        return checkNotNull(stream, "stream").getProperty(stateKey);
+    }
+
+    private class EnqueueStrategy {
```
If we only envision there being these 2 strategies can we consider removing this class?  We could consolidate the usage of `state.enqueued`, remove a few types, and remove the ` EnqueueStrategy enqueueStrategy` member for each `State` object:

```java
private final class State {
  private boolean hasBeenEnqueued;
  void enqueue() {
    if (hasBeenEnqueued) {
       queue.addLast(this);
    } else {
       queue.addFirst(this);
       hasBeenEnqueued = true;
    }
    enqueued = true;
  }

  static State dequeue() {
    State s = queue.removeFirst();
    s.enqueued = false;
  }
}
```

nmittler replied:
I was trying to avoid the extra conditional check on every call to `enqueue`.

Scottmitch replied:
ok thanks fine as is

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,252 @@
+                streamableBytes = newStreamableBytes;
+                totalStreamableBytes += delta;
+
+                if (!enqueued && streamableBytes > 0) {
```
Do we need to handle "empty" frames here? Maybe take into account `hasFrame`?

nmittler replied:
I don't think it matters for allocation.

Scottmitch replied:
I see. Same as the priority distributor you are always allocating over all active streams for the write.  I guess we could potentially limit this to what is in the queue if we include EMPTY elements ... but doesn't seem necessary now.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,252 @@
+        Writer writer;
+
+        void writeAllocatedBytes(Writer writer) throws Http2Exception {
+            this.writer = writer;
```
Should we add re-entry protection like in `PriorityStreamByteDistributor`?

Also it seems like this `WriteVisitor` and `PriorityStreamByteDistributor.WriteVisitor` are doing the same thing. Is it worth sharing code with a package private class, considering the `State` type is different?

nmittler replied:
done.  It's not much code to share ... and sharing it may be awkward.  I'm fine with a bit of copy & paste here.

Scottmitch replied:
sgtm

Scottmitch commented on microbench/src/main/java/io/netty/microbench/http2/NoPriorityByteDistributionBenchmark.java:
```diff
@@ -0,0 +1,315 @@
+    @Param({ "100", "10000" })
+    private int numStreams;
+
+    @Param({"1024", "65536" /*, "1048576"*/})
```
`1048576` ... anything interesting to report here?

nmittler replied:
Just didn't feel like waiting for results :)

Scottmitch commented on microbench/src/main/java/io/netty/microbench/http2/NoPriorityByteDistributionBenchmark.java:
```diff
@@ -0,0 +1,315 @@
+public class NoPriorityByteDistributionBenchmark extends AbstractMicrobenchmark {
+    public Http2Connection connection;
+
+    public enum Algorithm {
```
`private`?

nmittler replied:
The enum has to be public for JMH to use it.

Scottmitch commented on microbench/src/main/java/io/netty/microbench/http2/NoPriorityByteDistributionBenchmark.java:
```diff
@@ -0,0 +1,315 @@
+@Fork(1)
+@State(Scope.Benchmark)
+public class NoPriorityByteDistributionBenchmark extends AbstractMicrobenchmark {
+    public Http2Connection connection;
```
`private`?

nmittler replied:
done.

Scottmitch commented:
@nmittler - Finished first pass. Looks good! Can we tie this distributor into unit tests? Particularly into tests with EMPTY frames?

nmittler commented:
@normanmaurer @Scottmitch PTAL

Scottmitch commented on microbench/src/main/java/io/netty/microbench/http2/NoPriorityByteDistributionBenchmark.java:
```diff
@@ -0,0 +1,315 @@
+    @Param({ "100", "10000" })
+    private int numStreams;
+
+    @Param({ "1024", "65536"/*, "1048576"*/ })
```
can you report results with `1048576` after updates are made? I expect them to be good, but better to be sure...

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,255 @@
+ * A {@link StreamByteDistributor} that ignores stream priority and uniformly allocates bytes
+ * to all streams.
+ */
+public class UniformStreamByteDistributor implements StreamByteDistributor {
```
final ?

nmittler replied:
done.

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,255 @@
+        this(connection, DEFAULT_MIN_ALLOCATION_CHUNK);
+    }
+
+    public UniformStreamByteDistributor(Http2Connection connection, int minAllocationChunk) {
```
maybe adding some javadocs and explain what the `minAllocationChunk` is ?

nmittler replied:
done.

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,255 @@
+        }
+
+        long allocated = 0;
+        int allocationChunk = max(minAllocationChunk, (int) (totalStreamableBytes / queue.size()));
```
what about:

```java
    int size = queue.size();
    if (size == 0 || maxBytes <= 0) {
            return 0L;
        }

        long allocated = 0;
        int allocationChunk = max(minAllocationChunk, (int) (totalStreamableBytes / size));
```

This way we not need to call `isEmpty()` and `size()`

nmittler replied:
done.

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,255 @@
+
+        long allocated = 0;
+        int allocationChunk = max(minAllocationChunk, (int) (totalStreamableBytes / queue.size()));
+        while (maxBytes > 0 && !queue.isEmpty()) {
```
Do the change mention above and change to : `while (maxBytes > 0 && --size >= 0)`

nmittler replied:
done.

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,255 @@
+            // now on.
+            state.enqueueStrategy = defaultEnqueueStrategy;
+        }
+    };
```
maybe add both "final" fields to the top of the class so these are grouped with the others ?

nmittler replied:
done.

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,255 @@
+        EnqueueStrategy enqueueStrategy = firstEnqueueStrategy;
+
+        State() {
+        }
```
remove constructor ?

nmittler replied:
done.

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,255 @@
+    /**
+     * A connection stream visitor that delegates to the user provided visitor.
+     */
+    private class WriteVisitor implements Http2StreamVisitor {
```
final ?

nmittler replied:
done.

nmittler commented:
@Scottmitch @normanmaurer PTAL

nmittler commented:
@Scottmitch @ejona86 @normanmaurer I've updated the benchmarks with additional counters to keep track of: minimum bytes written per stream, total written per stream, and number of streams written (all per invocation averages).

It's pretty clear now that the per-stream goodput blows away the priority tree algorithm.

normanmaurer commented:
@nmittler ship it dude!

nmittler commented:
@Scottmitch any other comments before I cherry-pick?

ejona86 commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,266 @@
+
+    public UniformStreamByteDistributor(Http2Connection connection, int minAllocationChunk) {
+        this.connection = checkNotNull(connection, "connection");
+        if (minAllocationChunk <= 0) {
```
<= 1

A `minAllocationChunk` of `0` could allow no streams to make progress (although you would need a _lot_ of streams).

nmittler replied:
Err ... I'm not allowing `0`. Or am I missing something?

ejona86 replied:
Doh. Yeah, it's fine.

ejona86 commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,266 @@
+        long unallocatedStreamableBytes = totalStreamableBytes - allocated;
+
+        // Need to write even if maxBytes == 0 in order to handle the case of empty frames.
+        writeVisitor.writeAllocatedBytes(writer);
```
So part of the point of this algorithm is that it is greedy and doesn't need this second pass. Do we get much benefit from not writing immediately? I would expect writing out immediately to produce simpler code (less book keeping), be more fair, and be faster (since it doesn't have to loop through all streams).

I would expect the main argument to be to allow write combining for the same stream when it is allocated bytes multiple times. However, I could easily be in favor of _not_ write combining, because write combining is unfair from a queuing standpoint. Any writes that occur here go into another queue to be written. Since we are putting an upper bound on the outbound queue this unfairness is not too big of a deal. But the point is that write combining isn't all that important, so we can take it or leave it.

Note that if you don't have something like `minAllocationChunk`, then write combining is important in order to prevent writing one byte at a time. But since this does have `minAllocationChunk`, `minAllocationChunk` is already protecting against "unreasonably small" writes.

nmittler replied:
Possibly, but if we don't combine writes we will be sending more data per frame (extra framing overhead). I'm not sure taking one extra pass at the end is going to be a significant hit in performance.

Scottmitch replied:
I was originally assuming we were allocating and writing in one pass and was worried about empty frames... However I think we can proceed the way the PR is now and improve later (include streams with empty frames in the queue, and allocate/write in a single pass) if necessary.  @ejona86 @nmittler - WDYT?

ejona86 replied:
If you have 100 idle streams, I would expect a noticeable difference. Not having to loop over all streams was a big reason I was wanting to change the algorithm. Any scheduling algorithm that has to loop over all items when choosing the next "is doing it wrong."

It is true that we can reduce number of data frames by write combining. However, the gains are expected to be small. If your `minAllocationChunk` is 1K, then the 9 byte frame overhead is < 1%. If the `minAllocationChunk` is 16K then the frame overhead is ~0.05%. Yes, you can have lower overhead by doing more combining, but it simply stops mattering at a point. Also, the cases you want a small `minAllocationChunk` are also the cases you don't want to combine the writes, because the link is slow and you don't want to monopolize the connection with a single stream as long.

@Scottmitch, I'm okay with committing and iterating.

nmittler replied:
@ejona86 OK that's fair. I've reworked the code a bit so that we only write to those streams that have data to send.  PTAL.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,266 @@
+     * The remote flow control state for a single stream.
+     */
+    private final class State {
+        boolean hasFrame;
```
is this used for anything?

nmittler replied:
no ... removed.

Scottmitch commented on codec-http2/src/test/java/io/netty/handler/codec/http2/UniformStreamByteDistributorTest.java:
```diff
@@ -0,0 +1,306 @@
+        List<Integer> streamsWrittenZeroBytes = new ArrayList<Integer>();
+        int bytesWritten = 0;
+        int streamsWritten = 0;
+        for (int ix = 0; ix < streamsToCapture.length; ++ix) {
```
nit: `ix` -> `i`?

nmittler replied:
this method has been deleted.

Scottmitch commented on codec-http2/src/test/java/io/netty/handler/codec/http2/UniformStreamByteDistributorTest.java:
```diff
@@ -0,0 +1,306 @@
+        int written = 3 * chunkSize;
+        assertTrue(write(written));
+        List<Integer> streamsNotWritten = verifyWrite(written, 3, STREAM_A, STREAM_B, STREAM_C, STREAM_D);
+        assertEquals(1, streamsNotWritten.size());
```
should we also assert which stream we expect not to be written?  is it `stream_a`?

nmittler replied:
Oops ... I meant to delete this test.

Scottmitch commented:
@nmittler - LGTM! Thanks for the updated benchmarks and tests!

ejona86 commented:
@nmittler, LGTM, and we can iterate.

nmittler commented:
@ejona86 @Scottmitch @normanmaurer I've reworked this so that we don't bother combining writes. This simplifies the code a bit.

I also removed the use of Mockito in the benchmarks as this was showing up as a hotspot in the profiler and was corrupting the benchmarks.  The updated benchmark results show that the new algorithm blows away the priority algorithm on all fronts (including throughput).

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,203 @@
+    private final class State {
+        final Http2Stream stream;
+        int streamableBytes;
+        boolean hasFrame;
```
Not to sound like a broken record, but is this used :smile: ?

nmittler replied:
Yes, it is used now.  It's what controls when a frame is enqueued.

ejona86 replied:
It looks like it uses a local variable, not this field.

nmittler replied:
Done.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,203 @@
+
+        // Write until the queue is empty or we've exhausted maxBytes. We need to check queue.isEmpty()
+        // here since the Writer could call updateStreamableBytes, which may alter the queue.
+        while (maxBytes > 0 && !queue.isEmpty()) {
```
You have already checked empty and max bytes in the `if` conditional above. Consider making this a `do {} while();`

nmittler replied:
Now that I'm using an iterator, this is changed to a for loop and I believe your concern is addressed.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,203 @@
+        checkNotNull(writer, "writer");
+
+        final int size = queue.size();
+        if (size == 0 || maxBytes <= 0) {
```
what if `maxBytes == 0` and some streams have empty frames to send?

nmittler replied:
Done. Using an index to make sure we iterate over all of the streams at least once.

ejona86 replied:
It seems @Scottmitch's case is still an issue, because this method will return before getting to the for loop.

It seems "stream has empty frame to send" is continually a problem. Why don't we just send immediately instead of enqueuing in that case?

Draining and refilling the queue to find empty streams seems like an overkill. If there turns out to be some reason we can't just send such things immediately, then I propose two alternatives:

 1. Keep a separate queue of streams that only have empty frames to send, and always drain that queue in `distribute()`.
 2. Just wait until you get to such streams in the queue:

    ```java
    while (!queue.isEmpty()) {
      State state = queue.remove();
      if (!(maxBytes > 0 || state.streamableBytes == 0)) {
        break;
      }
      ...
    }
    ```

nmittler replied:
oof, good catch ... needed more coffee :)

The way things are currently structured, we can't just write empty frames immediately, because the writer is provided in `distribute`. Perhaps we should just provide a setter for the writer instead? @Scottmitch thoughts?

I think the easiest solution for now is to just use a second queue as you proposed. We can't rely on iterating past the last zero-length frame because new frames are also added to the front of the queue, so zero-length frames could be interspersed with non-zero-length frames.

nmittler commented:
@Scottmitch PTAL

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,211 @@
+        // here since the Writer could call updateStreamableBytes, which may alter the queue.
+        // We iterate through all streams at least once in order to make sure that any zero-length
+        // frames are written.
+        for (int numWritten = 0; numWritten < size || (maxBytes > 0 && !queue.isEmpty()); ++numWritten) {
```
do we need the queue.isEmpty() call ? Why not just use the size and doing some decrementation ?

nmittler replied:
The comment above explains .. basically the fact that we're calling `Writer.write` within the loop implies that the queue can be altered externally.

normanmaurer replied:
ah I see.. thanks

normanmaurer commented on codec-http2/src/test/java/io/netty/handler/codec/http2/UniformStreamByteDistributorTest.java:
```diff
@@ -0,0 +1,236 @@
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
```
remove empty line

nmittler replied:
done.

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,211 @@
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
```
remove empty line

nmittler replied:
done.

nmittler commented:
@Scottmitch @ejona86 PTAL

ejona86 commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,203 @@
+         * assuming all of the bytes will be written.
+         */
+        void write(int numBytes, Writer writer) throws Http2Exception {
+            if (numBytes == 0 && streamableBytes > 0) {
```
I think this condition is now unnecessary/never triggered. The callers are now guaranteeing `numBytes > 0 || streamableBytes == 0`

ejona86 commented:
@nmittler LGTM

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,203 @@
+
+        // First, write out any empty frames.
+        while (!emptyFrameQueue.isEmpty()) {
+            State state = emptyFrameQueue.remove();
```
I'm in favor of keeping 1 queue and sending empty frames immediately (if possible, i.e. nothing is already queued).  Keeping 2 queues seems to be more complicated and it can be confusing that state in the `queue` may also have empty frames to send. These 2 different queues have different ordering semantics to conceptualize and coordinate.  For example what happens if a `State` object initially has an empty frame, and then a non-empty frame before a `write` event takes place? It seems like we use `enqueued` to indicate the `State` object is in `emptyFrameQueue` or `queue`, and then enforce that we can only add the `State` object to 1 of the queues (not both).

nmittler replied:
@Scottmitch keeping them all into a single queue requires iteration across all streams, which is a fairly significant cost. Treating empty frames as a special case and out of the main queue has significant performance gains that are measurable by this benchmark.

The 2 queue approach isn't really all that complicated and the new code handles the case of migration of an empty frame to a non-empty frame fairly easily.

Scottmitch replied:
@nmittler - With the current 2 queue approach is seems like we only get the benefit of not having to iterate through the entire queue is if for a given stream the `streamableBytes == 0`. This condition will be false the first time a "non-empty" stream has been added since the last write operation. What if a stream has the following frames sizes: `1024, 0` before a write operation?  In this case is it possible we will service the first 1024 byte frame, and not service the 0 byte frame (assume the two frames cannot be merged, e.g. data and trailing headers)?

I wonder if we could simplify, and have a lower memory footprint, by just using 1 queue and either always iterate through it (what is the perf implications), or maybe tracking a `boolean queueHasEmptyFrame` to be able to short circuit the full iteration in some situations. WDYT?

nmittler replied:
@Scottmitch Just to be clear, let me explain the performance advantage of the 2 queue approach ...

Assume that we have a flow control window of 1K and 1000 streams, each with > 1K to write. This is a pretty common case. If we have zero-length frames in the main queue, we would have to iterate over all the streams which would be O(n).  If we can separate the zero-length frames into a separate queue, we would write our 1K to the first stream and be done, which would be O(1). This advantage shows up clearly in our benchmarks. I had tried the O(n) approach and the throughput dropped significantly.

> What if a stream has the following frames sizes: `1024, 0` before a write operation? In this case is it possible we will service the first 1024 byte frame, and not service the 0 byte frame (assume the two frames cannot be merged, e.g. data and trailing headers)?

From the perspective of the distributor, it doesn't matter. It will distribute 1024 to the `Writer` callback. The flow controller will iterate over the 2 flow controlled objects and write them both.

> I wonder if we could simplify, and have a lower memory footprint,...

The additional memory will be minimal as we don't expect a lot of zero-length frames to be added. If for some strange reason the user start writing initial frames that are empty, the worst case is that we'll have 2 queues of length n, rather than one.  Since this case isn't all that bad or likely, I think the tradeoff in throughput far outweights this possibility of additional memory.

Scottmitch replied:
> Just to be clear, let me explain the performance advantage of the 2 queue approach ...

Yip benefits are understood :) I didn't notice that while iterating the `emptyFrameQueue` you were checking `state.streamableBytes > 0` and then appending to `queue` instead of writing immediately (did this change recently?).

> The additional memory will be minimal as we don't expect a lot of zero-length frames to be added

Aren't header frames zero-length?  In some cases (when headers and data are written at same time) this may mean we are actually doing 2 iterations over collections of size `n`.

-> talked offline with @nmittler - we should not put headers through the flow controller.

Scottmitch commented on microbench/src/main/java/io/netty/microbench/http2/NoPriorityByteDistributionBenchmark.java:
```diff
@@ -0,0 +1,620 @@
+    }
+
+    /*
+    public static void main(String[] args) throws Exception {
```
remove?

nmittler replied:
done.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,203 @@
+            }
+        }
+
+        void enqueue() {
```
private?

nmittler replied:
Not sure why ... the class is already private and it's called from the main class.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,203 @@
+        boolean enqueued;
+        boolean wasEnqueued;
+
+        private State(Http2Stream stream) {
```
package private?

nmittler replied:
done.

nmittler commented:
@normanmaurer @Scottmitch @ejona86 PTAL

I've simplified the benchmark class and gave it better counters to make the picture more clear.

I also fixed a bug in the flow controller where it was updating stream state with streamable bytes but no frame.

nmittler commented:
FYI, there are some other bugs in `DefaultHttp2RemoteFlowController` that I'm tracking down which are causing the test failures.  Stay tuned...

nmittler commented:
@Scottmitch @ejona86 I've found a few bugs in `DefaultHttp2RemoteFlowController` that are blocking this issue and have created PR #4483 to fix it.  Once that's in I'll rebase and ping you for another look.

nmittler commented:
@Scottmitch @ejona86 I've rebased on top of the flow controller fix.  PTAL

ejona86 commented:
@nmittler, still LGTM

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,210 @@
+
+    private final Http2Connection.PropertyKey stateKey;
+    private final Deque<State> queue = new ArrayDeque<State>();
+    private final Deque<State> emptyFrameQueue = new ArrayDeque<State>();
```
Can we lazily initialize this queue, and initialize the size to something smaller? When we make the change so that headers are no longer queued in the flow controller (https://github.com/netty/netty/issues/4489) this will be a good thing to have.

nmittler replied:
done.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,210 @@
+    static final int DEFAULT_MIN_ALLOCATION_CHUNK = 1024;
+
+    private final Http2Connection.PropertyKey stateKey;
+    private final Deque<State> queue = new ArrayDeque<State>();
```
Should we add a constructor which accepts an initialize size for this queue?

nmittler replied:
I just checked for initial capacities that we're using elsewhere.  In `DefaultHttp2Connection`, we specify `4` as the default capacity.  Let's just go with that and we'll use `2` for the empty queue.  WDYT?

nmittler commented:
@Scottmitch PTAL

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,210 @@
+                    queue.addLast(this);
+                }
+                enqueued = true;
+                wasEnqueued = true;
```
if the item is added to `emptyFrameQueue` but then more frames are added we will remove it from `emptyFrameQueue` and add it to the end of `queue`. Should we just add this element to the front of `queue`? Should we move this `wasEnqueued = true` when we first add to `wasEnqueued` (and maybe rename)?

nmittler replied:
done.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java:
```diff
@@ -0,0 +1,219 @@
+    static final int DEFAULT_MIN_ALLOCATION_CHUNK = 1024;
+
+    private final Http2Connection.PropertyKey stateKey;
+    private final Deque<State> queue = new ArrayDeque<State>(4);
```
seems like some users may want this to be bigger by default? For example all data frames will go through this queue. If we are sending with high frequency this queue may have to re-size/copy a few times before getting up to equilibrium.  Maybe if the user doesn't specify then we can pick a smaller number, but it may still be good to allow users to have a constructor arg to control this.

nmittler replied:
This queue will only ever be as big as the number of active streams. So the same would be true for `DefaultHttp2Connection`.  I don't think we should handle this structure differently than other parts of the code.  If we don't like this, we should address it separately ... and apply the change across the codebase.  WDYT?

Scottmitch replied:
thanks for clarifying... +1 to number of active streams.  agreed that we should be consistent. fine as is.

Scottmitch commented:
@nmittler - Good work! ship it!

normanmaurer commented:
@nmittler @Scottmitch :beers:

nmittler commented:
Thanks for the review guys!

Cherry-picked into master (ebfe3d2) and 4.1 (2a2059d)
